File: Graph/13549번 - 숨바꼭질 3/13549.cpp
#include <iostream>
#include <deque>

using namespace std;

#define X first
#define Y second

int n, k;
int vis[100001];

void input() {
    cin >> n >> k;
    fill_n(&vis[0], 100001, -1);
}

void solve() {
    deque<int> dq;
    dq.push_front(n);
    vis[n] = 0;
    
    while(!dq.empty()) {
        int cur = dq.front();
        dq.pop_front();
        
        for(int nxt : {cur * 2, cur - 1, cur + 1}) {
            if(nxt < 0 || nxt >= 100001 || vis[nxt] != -1) continue;
            
            if(nxt == cur * 2) {
                vis[nxt] = vis[cur];
                dq.push_front(nxt);
            }
            else {
                vis[nxt] = vis[cur] + 1;
                dq.push_back(nxt);
            }
        }
    }
    
    cout << vis[k] << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Graph/13549번 - 숨바꼭질 3/13549.py
import sys
from collections import deque

input = sys.stdin.readline

LIM = 100001

def solve(n, k):
    dq = deque()
    dq.append(n)
    
    vis = [-1] * LIM
    vis[n] = 0
    
    while dq:
        cur = dq.popleft()
        
        if cur == k:
            print(vis[cur])
        
        for nxt in [cur - 1, cur + 1, cur * 2]:
            if (nxt < 0 or nxt >= LIM) or vis[nxt] != -1:
                continue
            
            if nxt == cur * 2:
                vis[nxt] = vis[cur]
                dq.appendleft(nxt)
            else:
                vis[nxt] = vis[cur] + 1
                dq.append(nxt)
                
    

if __name__ == '__main__':
    n, k = map(int, input().split())
    solve(n, k)